client.go

```diff
@@ -0,0 +1,268 @@
+
+	"github.com/bufbuild/connect-go"
+	reflectionv1 "github.com/bufbuild/connect-grpcreflect-go/internal/gen/go/connectext/grpc/reflection/v1"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func NewClientV1(httpClient connect.HTTPClient, baseURL string, options ...connect.ClientOption) *Client {
+	return newClient(
+		httpClient,
+		baseURL+serviceNameV1,
+		options...,
+	)
+}
+
+func NewClientV1Alpha(httpClient connect.HTTPClient, baseURL string, options ...connect.ClientOption) *Client {
+	return newClient(
+		httpClient,
+		baseURL+serviceNameV1Alpha,
+		options...,
+	)
+}
+
+func newClient(httpClient connect.HTTPClient, serviceURL string, options ...connect.ClientOption) *Client {
+	client := connect.NewClient[reflectionv1.ServerReflectionRequest, reflectionv1.ServerReflectionResponse](
+		httpClient,
+		serviceURL+methodName,
+		options...,
+	)
+	return &Client{client: client}
+}
+
+type reflectClient = connect.Client[reflectionv1.ServerReflectionRequest, reflectionv1.ServerReflectionResponse]
+type reflectStream = connect.BidiStreamForClient[reflectionv1.ServerReflectionRequest, reflectionv1.ServerReflectionResponse]
+
+type Client struct {
+	client *reflectClient
+}
+
+type ClientStreamOption func(*clientStreamOptions)
+
+func WithRequestHeaders(headers http.Header) ClientStreamOption {
+	return func(o *clientStreamOptions) {
+		o.headers = headers
+	}
+}
+
+func WithReflectionHost(host string) ClientStreamOption {
+	return func(o *clientStreamOptions) {
+		o.host = host
+	}
+}
+
+type clientStreamOptions struct {
+	host    string
+	headers http.Header
+}
+
+func (c *Client) NewReflectionStream(ctx context.Context, options ...ClientStreamOption) *ClientStream {
+	var opts clientStreamOptions
+	for _, option := range options {
+		option(&opts)
+	}
+	stream := c.client.CallBidiStream(ctx)
+	for k, v := range opts.headers {
+		stream.RequestHeader()[k] = v
+	}
+	// we can eagerly send request headers; we can ignore return
+	// value because caller will see any errors when calling any
+	// other method on returned stream
+	_ = stream.Send(nil)
+	return &ClientStream{host: opts.host, stream: stream}
+}
+
+func IsReflectionStreamError(err error) bool {
+	var streamErr *streamError
+	return errors.As(err, &streamErr)
+}
+
+type ClientStream struct {
+	host string
+
+	mu     sync.Mutex
+	stream *reflectStream
+}
+
+func (cs *ClientStream) Peer() connect.Peer {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	return cs.stream.Peer()
+}
+
+func (cs *ClientStream) ResponseHeader() http.Header {
+	cs.mu.Lock()
+	stream := cs.stream
+	cs.mu.Unlock()
+	// this will block until headers received, so we can't hold lock
+	// while calling or else we'll potentially deadlock calls to send
+	return stream.ResponseHeader()
+}
+
+func (cs *ClientStream) ResponseTrailer() http.Header {
+	cs.mu.Lock()
+	stream := cs.stream
+	cs.mu.Unlock()
+	return stream.ResponseTrailer()
+}
```

Comment:
We're hiding `CloseRequest` and `CloseResponse`, so there's no chance that the stream is half-closed. Would it be less error-prone if we returned the trailers from `Close`, so users get a stronger signal of when it's safe to inspect them?

Reply:
I think that makes sense. Done.